src/V3HierBlock.cpp

```diff
@@ -0,0 +1,420 @@
+    opts.push_back(" --hierarchical-child");
+
+    const StrGParams gparamsStr = stringifyParams(gparams(), true);
+    for (StrGParams::const_iterator paramIt = gparamsStr.begin(); paramIt != gparamsStr.end();
+         ++paramIt) {
+        opts.push_back("-G" + paramIt->first + "=" + paramIt->second + "");
+    }
+    return opts;
+}
+
+V3StringList V3HierBlock::hierBlockOptions() const {
+    V3StringList opts;
+    const StrGParams gparamsStr = stringifyParams(gparams(), false);
+    opts.push_back("--hierarchy-block ");
+    string s = modp()->origName();  // origName
+    s += "," + modp()->name();  // mangledName
+    for (StrGParams::const_iterator paramIt = gparamsStr.begin(); paramIt != gparamsStr.end();
+         ++paramIt) {
+        s += "," + paramIt->first;
+        s += "," + paramIt->second;
+    }
+    opts.back() += s;
+    return opts;
+}
+
+string V3HierBlock::hierPrefix() const { return "V" + modp()->name(); }
+
+string V3HierBlock::hierSomeFile(bool withDir, const char* prefix, const char* suffix) const {
+    string s;
+    if (withDir) s = hierPrefix() + '/';
+    s += prefix + modp()->name() + suffix;
+    return s;
+}
+
+string V3HierBlock::hierWrapper(bool withDir) const { return hierSomeFile(withDir, "", ".sv"); }
+
+string V3HierBlock::hierMk(bool withDir) const { return hierSomeFile(withDir, "V", ".mk"); }
+
+string V3HierBlock::hierLib(bool withDir) const { return hierSomeFile(withDir, "lib", ".a"); }
+
+string V3HierBlock::hierGenerated(bool withDir) const {
+    return hierWrapper(withDir) + ' ' + hierMk(withDir);
+}
+
+void V3HierBlock::writeCommandFile(bool forCMake) const {
+    vl_unique_ptr<std::ofstream> of(V3File::new_ofstream(commandFileName(forCMake)));
+    *of << "--cc\n";
+
+    if (!forCMake) {
+        for (V3HierBlock::HierBlockSet::const_iterator child = m_children.begin();
+             child != m_children.end(); ++child) {
+            *of << v3Global.opt.makeDir() << "/" << (*child)->hierWrapper(true) << "\n";
+        }
+    }
+    *of << "-Mdir " << v3Global.opt.makeDir() << "/" << hierPrefix() << " \n";
+    writeCommonInputs(of.get(), forCMake);
+    const V3StringList& commandOpts = commandOptions(false);
+    for (V3StringList::const_iterator it = commandOpts.begin(); it != commandOpts.end(); ++it) {
+        *of << (*it) << "\n";
+    }
+    *of << hierBlockOptions().front() << "\n";
+    for (HierBlockSet::const_iterator child = m_children.begin(); child != m_children.end();
+         ++child) {
+        *of << hierBlockOptions().front() << "\n";
+    }
+    *of << v3Global.opt.allArgsStringForHierBlock(false) << "\n";
+}
+
+string V3HierBlock::commandFileName(bool forCMake) const {
+    return hierCommandFileName(hierPrefix(), forCMake);
+}
+
+//######################################################################
+// Collect how hierarchy blocks are used
+class HierBlockUsageCollectVisitor : public AstNVisitor {
```

Comment:
Consider if cleaner if this visitor is a separate C++ file that includes the hier block header?

Reply:
I'm not sure if moving the class to a new .cpp makes the code cleaner.
The visitor class is a kind of private member of V3HierBlockPlan and not useful for other classes.
It is about 70 lines out of 436 lines of V3HierBlock.cpp.
436 lines does not look too big in the Verilator codebase.

Reply:
That's fine.